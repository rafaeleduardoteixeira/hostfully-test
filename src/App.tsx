import React from 'react';
import { createBrowserRouter, RouterProvider } from 'react-router-dom';
import createdRoutes from './constants/router';
import { Header } from './components/Header/Header';
import { AppContainer } from './App.styles';

function App() {
  return (
    <AppContainer>
      <Header />
      <RouterProvider router={createdRoutes} />
    </AppContainer>
  );
}

export default App;
