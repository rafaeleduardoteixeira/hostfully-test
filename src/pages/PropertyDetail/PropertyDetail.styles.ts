import styled from 'styled-components';
import { colors, media } from '../../themes';

export const Container = styled.div`
  display: flex;
  flex-direction: column;
  gap: 20px;

  padding-bottom: 20%;
`;

export const PropertyImageContainer = styled.div`
  width: 100%;
  height: 400px;
  background-color: ${colors.black};
  position: relative;

  @media ${media.mobile} {
    display: flex;
    flex-direction: column;
    justify-content: center;
    gap: 20px;
  }
`;

export const PropertyImageBanner = styled.img`
  opacity: 0.8;
  width: 100%;
  height: 100%;
  object-fit: cover;
  position: absolute;
`;

export const PropertyNameBanner = styled.h1`
  display: flex;
  color: white;
  font-size: 24px;
  font-weight: 700;
  width: 100%;
  bottom: 20px;
  position: absolute;
  padding: 0 20px;

  @media ${media.mobile} {
    z-index: 2;
    position: unset;
  }
`;

export const PropertyImageButtonBanner = styled.button`
  position: absolute;
  bottom: 20px;
  right: 20px;
  background-color: white;
  border: none;
  padding: 10px;
  border-radius: 10px;
  cursor: pointer;
  font-size: 16px;
  font-weight: 700;
  color: ${colors.black};

  @media ${media.mobile} {
    z-index: 2;
    align-self: center;
    position: unset;
  }
`;

export const PropertyImagemContainer = styled.div`
  display: flex;
  gap: 20px;
  overflow-y: scroll;

  @media ${media.mobile} {
    flex-direction: column;
  }
`;

export const PropertyImage = styled.img`
  width: 100%;
  height: 100%;
  object-fit: cover;
  border-radius: 10px;
`;

export const PropertyTopContent = styled.div`
  display: flex;
  gap: 20px;
  width: 100%;

  @media ${media.mobile} {
    flex-direction: column-reverse;
  }
`;

export const PropertyReturn = styled.img`
  cursor: pointer;
  width: 30px;
  height: 30px;
  z-index: 2;
`;
