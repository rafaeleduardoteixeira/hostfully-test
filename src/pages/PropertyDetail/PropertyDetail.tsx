import React, { useContext, useState } from 'react';
import { useNavigate, useParams } from 'react-router-dom';
import { AppContext, AppContextType } from '../../Context/App.context';
import {
  Container,
  PropertyImageContainer,
  PropertyImageBanner,
  PropertyNameBanner,
  PropertyImageButtonBanner,
  PropertyImage,
  PropertyImagemContainer,
  PropertyTopContent,
  PropertyReturn,
} from './PropertyDetail.styles';
import { Modal } from '../../components/Modal/Modal';
import { PropertyBookNow } from './components/PropertyBookNow/PropertyBookNow';
import { PropertyMap } from './components/PropertyMap/PropertyMap';
import { PropertyDescription } from './components/PropertyDescription/PropertyDescription';

export const PropertyDetail = () => {
  const { id } = useParams();
  const { getPropertyById } = useContext(AppContext) as AppContextType;
  const property = getPropertyById(id || '');
  const navigate = useNavigate();

  const [showAllImages, setShowAllImages] = useState(false);

  const handleShowAllImages = () => {
    setShowAllImages(true);
  };

  const handleCloseAllImages = () => {
    setShowAllImages(false);
  };

  const handleReturn = () => {
    navigate('/');
  };

  return (
    <Container data-testid="PropertyDetail">
      {property && (
        <>
          <PropertyReturn src="/icons/returnArrow.svg" onClick={handleReturn} />
          <PropertyImageContainer>
            <PropertyImageBanner src={property?.images[0]} onClick={handleShowAllImages} />
            <PropertyNameBanner>{property.name}</PropertyNameBanner>
            <PropertyImageButtonBanner onClick={handleShowAllImages} data-testid="ShowAllImages">
              See all photos
            </PropertyImageButtonBanner>
          </PropertyImageContainer>

          <PropertyTopContent>
            {property.position.longitude && property.position.latitude && <PropertyMap property={property} />}
            <PropertyBookNow property={property} />
          </PropertyTopContent>
          <PropertyDescription property={property} />
        </>
      )}
      {/* Created a modal to show all images from the property. It's a generic modal that can be used in different places. */}
      {showAllImages && (
        <Modal onClose={handleCloseAllImages}>
          <PropertyImagemContainer>
            {property?.images.map((image, index) => <PropertyImage src={image} key={index} />)}
          </PropertyImagemContainer>
        </Modal>
      )}
    </Container>
  );
};
