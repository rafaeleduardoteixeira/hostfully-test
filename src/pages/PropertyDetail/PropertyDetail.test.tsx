import React, { act } from 'react';
import '@testing-library/jest-dom';
import { render, screen } from '@testing-library/react';
import { PropertyDetail } from './PropertyDetail';
import { AppContext } from '../../Context/App.context';

const mockedUsedNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
  ...(jest.requireActual('react-router-dom') as any),
  useNavigate: () => mockedUsedNavigate,
}));

jest.mock('react-leaflet', () => jest.fn());

describe('PropertyDetail', () => {
  test('Snapshot test', () => {
    const { asFragment } = render(
      <AppContext.Provider
        value={{
          isLoading: false,
          getProperties: () => {
            return [];
          },
          getPropertyById: jest.fn(),
          getPropertyRentedDates: jest.fn(),
          getRentedProperties: jest.fn(),
          handleAddRentals: jest.fn(),
          handleRemoveRentals: jest.fn(),
          handleEditRentals: jest.fn(),
        }}
      >
        <PropertyDetail />
      </AppContext.Provider>,
    );
    expect(asFragment()).toMatchSnapshot();
  });
  test('Should render PropertyDetail', () => {
    render(
      <AppContext.Provider
        value={{
          isLoading: false,
          getProperties: () => {
            return [];
          },
          getPropertyById: jest.fn(),
          getPropertyRentedDates: jest.fn(),
          getRentedProperties: jest.fn(),
          handleAddRentals: jest.fn(),
          handleRemoveRentals: jest.fn(),
          handleEditRentals: jest.fn(),
        }}
      >
        <PropertyDetail />
      </AppContext.Provider>,
    );
    const PropertyDetailElement = screen.getByTestId('PropertyDetail');
    expect(PropertyDetailElement).toBeInTheDocument();
  });
  test('Should show all images', () => {
    render(
      <AppContext.Provider
        value={{
          isLoading: false,
          getProperties: () => {
            return [];
          },
          getPropertyById: () => {
            return {
              id: '1',
              name: 'Name 1',
              description: 'Description 1',
              city: 'City 1',
              country: 'Country 1',
              purpose: 'Purpose 1',
              price: 1,
              images: ['image'],
              position: {
                latitude: 0,
                longitude: 0,
              },
            };
          },
          getPropertyRentedDates: jest.fn(),
          getRentedProperties: jest.fn(),
          handleAddRentals: jest.fn(),
          handleRemoveRentals: jest.fn(),
          handleEditRentals: jest.fn(),
        }}
      >
        <PropertyDetail />
      </AppContext.Provider>,
    );
    const showAllImages = screen.getByText('See all photos');
    act(() => {
      showAllImages.click();
    });
    const PropertyImagemContainer = screen.getByTestId('Modal');
    expect(PropertyImagemContainer).toBeInTheDocument();
  });
});
