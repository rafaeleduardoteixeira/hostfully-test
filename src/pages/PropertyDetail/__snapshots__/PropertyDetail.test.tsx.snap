// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PropertyDetail Snapshot test 1`] = `
<DocumentFragment>
  <div
    class="sc-braxZu cGBMNQ"
    data-testid="PropertyDetail"
  />
</DocumentFragment>
`;
